### PR TITLE
Use promise based API

### PR DESCRIPTION
**Pull request by kevva**
One thing I want to improve before merging this is the error reporting. Should at least include which file that failed and maybe there are other things worth including.

@sindresorhus @shinnn

### CODE REVIEW

File: test.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+test('optimize a file', async t => {
+	const buf = await fsP.readFile(path.join(__dirname, 'fixture.jpg'));
+	const files = await m('fixture.jpg', {use: imageminJpegtran});
+
+	t.is(files[0].path, null);
+	t.true(files[0].data.length < buf.length);
+	t.true(isJpg(files[0].data));
+});
+
+test('#buffer optimize a file', async t => {
```
Comment (sindresorhus) [inline test.js]: `#` intended?

File: readme.md [hunk 2]
```diff
@@ -13,117 +13,65 @@ $ npm install --save imagemin
...
+### imagemin(input, output, [options])
 
-Creates a new `Imagemin` instance.
+Returns a promise for an array of objects in the format `{data: Buffer, path: String}`.
 
-### .src(file)
+#### input
 
-Type: `array`, `buffer` or `string`
+Type: `string`, `array`
```
Comment (sindresorhus) [inline readme.md]: Should it really support both string and array?

File: readme.md [hunk 3]
```diff
@@ -13,117 +13,65 @@ $ npm install --save imagemin
...
+const imageminPngquant = require('imagemin-pngquant');
+
+imagemin('images/*.{jpg,png}', 'build/images', {
+	use: [
+		imageminMozjpeg,
+		imageminPngquant
+	]
+}).then(files => {
+	console.log(files);
+	//=> [{data: <Buffer 89 50 4e ...>, path: 'build/images/foo.jpg'}, ...]
```
Comment (sindresorhus) [inline readme.md]: `...` => `…` ?

**Comment by sindresorhus** [comment]
Looks very good!

**Comment by sindresorhus** [comment]
Maybe try integrating it with gulp-imagemin before releasing this to ensure the API works fine for consumers.

**Comment by shinnn** [comment]
LGTM. I had [another concern about performance](https://github.com/imagemin/imagemin-jpegtran/commit/f2b61fc446b4c3b14ad40852dea4c4785bc4c770#commitcomment-17287860) but I found it doesn't matter.

File: readme.md (hunk 2 above)
Comment (kevva) [inline readme.md]: Hmm, maybe only string is needed?

**Comment by kevva** [comment]
@shinnn, oh, that's great. Feels more stable not using `stdin/stdout` too.

File: readme.md (hunk 2 above)
Comment (sindresorhus) [inline readme.md]: No, I mean, only array.

**Pull request merged**